### PR TITLE
fix: use pwasm-utils 0.18 for more correct stack accounting

### PR DESCRIPTION
**Pull request by matklad**
This forward-ports the change from 1.23.1

We upgrade our version of pwasm-utils -- the old one severely
undercounted stack usage in some cases. As this is a protocol change, we
keep the old implementation around

Test Plan
=========

* added test for stack overflow behavior for the bad case
* added test for protocol upgrade related to stack overflow
* fuzzed new pwasm utils implementation locally, confirmed that it only
  differs for calls which stack overflow

### CODE REVIEW

File: core/primitives/src/version.rs [hunk 1]
```diff
@@ -128,6 +128,9 @@ pub enum ProtocolFeature {
     /// Make block producers produce chunks for the same block they would later produce to avoid
     /// network delays
     SynchronizeBlockChunkProduction,
+    /// Change the algorithm to count WASM stack usage to avoid undercounting in
+    /// some cases.
+    CorrectStackLimit,
```
Comment (nagisa) [inline core/primitives/src/version.rs:133]: if we ever need to correct the algorithm again, it'll be hard to come up with a name for it ^^

**Review by nagisa — APPROVED** [review]

File: core/primitives/src/version.rs (hunk 1 above)
Comment (matklad) [reply core/primitives/src/version.rs:133]: This name we can retroactively change in the future, as it is not serialized anywhere, so tried not to over-thing this )

File: core/primitives/src/version.rs [hunk 2]
```diff
@@ -181,7 +184,8 @@ impl ProtocolFeature {
             | ProtocolFeature::LimitContractFunctionsNumber
             | ProtocolFeature::BlockHeaderV3
             | ProtocolFeature::AliasValidatorSelectionAlgorithm => 49,
-            ProtocolFeature::SynchronizeBlockChunkProduction => 50,
+            ProtocolFeature::SynchronizeBlockChunkProduction
+            | ProtocolFeature::CorrectStackLimit => 50,
```
Comment (mm-near) [inline core/primitives/src/version.rs:188]: shouldn't we use a new protocol version for this? (or was it already deployed somehow?)

File: runtime/near-vm-runner/src/prepare.rs [hunk 3]
```diff
@@ -172,26 +197,175 @@ impl<'a> ContractModule<'a> {
...
+                        imported_mem_type = Some(memory_type);
+                        continue;
+                    }
+                    _ => continue,
+                };
+
+                let Type::Function(ref _func_ty) =
+                    types.get(*type_idx as usize).ok_or(PrepareError::Instantiate)?;
+
+                // TODO: Function type check with Env
```
Comment (mm-near) [inline runtime/near-vm-runner/src/prepare.rs:327]: optional: I'm not a fan of 'orphaned' TODOs ;-) we should either have a github/jira issue that is tracking it (if this is a lot more complex fix), or author name + additional info on why this is complex enough that we're not fixing it in this PR.

File: runtime/near-vm-runner/src/prepare.rs [hunk 4]
```diff
@@ -172,26 +197,175 @@ impl<'a> ContractModule<'a> {
...
+                /*
+
+                let ext_func = env
```
Comment (mm-near) [inline runtime/near-vm-runner/src/prepare.rs:330]: do you need this code?

**Review by mm-near — APPROVED** [review]

File: core/primitives/src/version.rs (hunk 2 above)
Comment (matklad) [reply core/primitives/src/version.rs:188]: Yeah, that was deployed as a part of 1.23.1: https://github.com/near/nearcore-ghsa-62f6-rgx9-xcw2/pull/1

File: core/primitives-core/src/config.rs [hunk 5]
```diff
@@ -82,6 +84,53 @@ pub struct VMLimitConfig {
     pub max_functions_number_per_contract: Option<u64>,
 }
 
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum StackLimiterVersion {
+    V0,
+    V1,
```
[the remark this answers is not in the document]
Comment (matklad) [reply core/primitives-core/src/config.rs:101]: Yeah, good call, added some comments.

File: runtime/near-vm-runner/src/prepare.rs [hunk 6]
```diff
@@ -23,18 +21,48 @@ pub(crate) const WASM_FEATURES: wasmparser::WasmFeatures = wasmparser::WasmFeatu
...
+    wasmparser::Validator::new()
+        .wasm_features(WASM_FEATURES)
+        .validate_all(original_code)
+        .map_err(|_| PrepareError::Deserialization)?;
+
+    match config.limit_config.stack_limiter_version {
+        // Support for old protocol versions, where we incorrectly didn't
+        // account for many locals of the same type.
+        //
+        // See `test_stack_instrumentation_protocol_upgrade` test.
```
Comment (matklad) [inline runtime/near-vm-runner/src/prepare.rs:45]: FWIW, there's a cool trick to make such links to tests checkable (ie, the test fails if it doesn't actually touch the relevant code): https://docs.rs/cov-mark/latest/cov_mark/.

I am still on the fence where maintaining such coverage marks helps overall.

File: runtime/near-vm-runner/src/prepare.rs (hunk 4 above)
Comment (matklad) [reply runtime/near-vm-runner/src/prepare.rs:334]: Yeah, I agree that it's better to just remove wishful-thinking comments from this module. In general, every time I open this file I want to refactor it. 

I'd rather not do that in the PR though, as the goal here is to change a bit of the behavior we want to change, and to hold everything else constant.

**Pull request merged**